File: solutions/Previous/221109/RescueBoat.java
import java.util.Arrays;

public class RescueBoat {
    public int solution(int[] weights, int limit) {
        Arrays.sort(weights);

        int answer = 0;
        int index = 0;

        for (int i = weights.length - 1; i >= index; i -= 1) {
            if (weights[i] + weights[index] <= limit) {
                index += 1;
                answer += 1;
                continue;
            }

            if (weights[i] + weights[index] > limit) {
                answer += 1;
            }
        }

        return answer;
    }
}

File: solutions/Previous/221110/rescueBoat.js
export default function solution(weights, limit) {
  const sortedWeights = weights.sort((a, b) => (b - a));

  let lightWeightIndex = sortedWeights.length - 1;

  const answer = sortedWeights.reduce((count, heavyWeight, index, array) => {
    if (index > lightWeightIndex) {
      return count;
    }
    if (heavyWeight + array[lightWeightIndex] <= limit) {
      lightWeightIndex -= 1;
      return count + 1;
    }
    if (index === lightWeightIndex
      || heavyWeight + array[lightWeightIndex] > limit) {
      return count + 1;
    }
    return count;
  }, 0);

  return answer;
}
